src/Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs

```diff
@@ -115,207 +90,123 @@ private void OnTransactionAvailable(TransactionReceived transactionReceived)
             this.ProcessTransaction(transactionReceived.ReceivedTransaction);
         }
 
-        private void OnBlockConnected(BlockConnected blockConnected)
-        {
-            this.ProcessBlock(blockConnected.ConnectedBlock.Block);
-        }
-
         /// <inheritdoc />
         public void Stop()
         {
-            this.signals.Unsubscribe(this.blockConnectedSubscription);
-            this.signals.Unsubscribe(this.transactionReceivedSubscription);
-        }
-
-        /// <summary>Called when a <see cref="Block"/> is added to the <see cref="blocksQueue"/>.
-        /// Depending on the <see cref="WalletTip"/> and incoming block height, this method will decide whether the <see cref="Block"/> will be processed by the <see cref="WalletManager"/>.
-        /// </summary>
-        /// <param name="block">Block to be processed.</param>
-        /// <param name="cancellationToken">The cancellation token.</param>
-        private async Task OnProcessBlockAsync(Block block, CancellationToken cancellationToken)
-        {
-            Guard.NotNull(block, nameof(block));
-
-            long currentBlockQueueSize = Interlocked.Add(ref this.blocksQueueSize, -block.BlockSize.Value);
-            this.logger.LogDebug("Queue sized changed to {0} bytes.", currentBlockQueueSize);
-
-            ChainedHeader newTip = this.chainIndexer.GetHeader(block.GetHash());
-
-            if (newTip == null)
-            {
-                this.logger.LogTrace("(-)[NEW_TIP_REORG]");
-                return;
-            }
-
-            // If the new block's previous hash is not the same as the one we have, there might have been a reorg.
-            // If the new block follows the previous one, just pass the block to the manager.
-            if (block.Header.HashPrevBlock != this.walletTip.HashBlock)
+            this.syncCancellationToken.Cancel();
+            while (this.walletSynchronisationLoop != null &&
+                   this.walletSynchronisationLoop.RunningTask.Status != TaskStatus.Canceled &&
+                   this.walletSynchronisationLoop.RunningTask.Status != TaskStatus.Faulted &&
+                   this.walletSynchronisationLoop.RunningTask.Status != TaskStatus.RanToCompletion)
```

Comment:
in my comment above I mean the whole block not just Thread.Yield

Reply:
Changed.